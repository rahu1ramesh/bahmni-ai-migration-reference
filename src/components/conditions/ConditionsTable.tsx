import React, { useMemo } from 'react';
import { Tag } from '@carbon/react';
import { ExpandableDataTable } from '@components/expandableDataTable/ExpandableDataTable';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { useConditions } from '@hooks/useConditions';
import { formatConditions } from '@services/conditionService';
import { ConditionStatus, FormattedCondition } from '@types/condition';
import { formatDate, formatDateTime } from '@utils/date';
import { FormatDateResult } from '@types/date';

/**
 * Maps condition status to appropriate tag type
 * @param status - The condition status
 * @returns The tag type for the status
 */
const getStatusTagType = (status: ConditionStatus): 'green' | 'gray' => {
  switch (status) {
    case ConditionStatus.Active:
      return 'green';
    case ConditionStatus.Inactive:
    default:
      return 'gray';
  }
};

/**
 * Component to display patient conditions in a DataTable with expandable rows
 */
const ConditionsTable: React.FC = () => {
  const patientUUID = usePatientUUID();
  const { conditions, loading, error } = useConditions(patientUUID);

  // Define table headers
  const headers = useMemo(
    () => [
      { key: 'display', header: 'Condition' },
      { key: 'status', header: 'Status' },
      { key: 'onsetDate', header: 'Onset Date' },
      { key: 'recorder', header: 'Provider' },
      { key: 'recordedDate', header: 'Recorded Date' },
    ],
    [],
  );

  // Format conditions for display
  const formattedConditions = useMemo(() => {
    if (!conditions || conditions.length === 0) return [];
    return formatConditions(conditions);
  }, [conditions]);

  // Function to render cell content based on the cell ID
  const renderCell = (condition: FormattedCondition, cellId: string) => {
    switch (cellId) {
      case 'display':
        return condition.display;
      case 'status':
        return (
          <Tag type={getStatusTagType(condition.status)}>
            {condition.status}
          </Tag>
        );
      case 'onsetDate': {
        const onsetDate: FormatDateResult = formatDate(
          condition.onsetDate || '',
        );
        return onsetDate.formattedResult || 'Not available';
      }
      case 'recorder':
        return condition.recorder || 'Not available';
      case 'recordedDate': {
        const recordedDate: FormatDateResult = formatDateTime(
          condition.recordedDate || '',
        );
        return recordedDate.formattedResult || 'Not available';
      }
    }
  };

  // Function to render expanded content for a condition
  const renderExpandedContent = (condition: FormattedCondition) => {
    if (condition.note && condition.note.length > 0) {
      return condition.note.map((note, index) => (
        <p style={{ padding: '0.5rem' }} key={index}>
          {note}
        </p>
      ));
    }
    return undefined;
  };

  return (
    <div
      style={{ width: '100%', paddingTop: '1rem' }}
      data-testid="condition-table"
    >
      <ExpandableDataTable
        tableTitle="Conditions"
        rows={formattedConditions}
        headers={headers}
        renderCell={renderCell}
        renderExpandedContent={renderExpandedContent}
        loading={loading}
        error={error}
        ariaLabel="Patient conditions"
        emptyStateMessage="No conditions found"
      />
    </div>
  );
};

export default ConditionsTable;
