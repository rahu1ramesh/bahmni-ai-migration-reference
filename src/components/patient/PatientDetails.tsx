import React, { useMemo } from 'react';
import { usePatient } from '@hooks/usePatient';
import { SkeletonText, Tile, Column, Grid } from '@carbon/react';
import { formatPatientData } from '@services/patientService';
import { usePatientUUID } from '@hooks/usePatientUUID';
import { Text } from '@carbon/react/lib/components/Text';

// TODO: Extract this as a PatientDetails Display Control Component
const PatientDetails: React.FC = () => {
  const patientUUID: string | null = usePatientUUID();
  const { patient, loading, error } = usePatient(patientUUID);

  // Format patient data using the service
  const formattedPatient = useMemo(() => {
    if (!patient) return null;
    return formatPatientData(patient);
  }, [patient]);

  if (loading || error || !patient || !formattedPatient) {
    return (
      <Tile>
        <SkeletonText
          heading
          width="100%"
          lineCount={5}
          data-testid="skeleton-loader"
        />
      </Tile>
    );
  }

  const formatField = (value?: string | number | null) =>
    value ? value : null;

  const formattedIdentifiers = formattedPatient.identifiers.size
    ? Array.from(formattedPatient.identifiers.entries())
        .map(([key, value]) => `${key}: ${value}`)
        .filter(Boolean)
        .join(' | ')
    : null;

  const formattedAge =
    formattedPatient.age && formattedPatient.age.years !== undefined
      ? `${formattedPatient.age.years} Years, ${formattedPatient.age.months} Months, ${formattedPatient.age.days} Days`
      : null;

  const details = [
    formatField(formattedPatient.gender),
    formattedAge,
    formatField(formattedPatient.birthDate),
  ]
    .filter(Boolean)
    .join(' | ');

  const contactInfo = [
    formatField(formattedPatient.formattedAddress),
    formatField(formattedPatient.formattedContact),
  ]
    .filter(Boolean)
    .join(' | ');

  return (
    <Tile>
      <Grid fullWidth>
        {/* Full Name as H2 */}
        <Column sm={4} md={8} lg={16}>
          {formattedPatient.fullName && <h2>{formattedPatient.fullName}</h2>}
        </Column>

        {/* Identifiers in single-line format */}
        {formattedIdentifiers && (
          <Column sm={4} md={8} lg={16}>
            <Text>{formattedIdentifiers}</Text>
          </Column>
        )}

        {/* Gender, Age, Birth Date in one line */}
        {details && (
          <Column sm={4} md={8} lg={16}>
            <Text style={{ textTransform: 'capitalize' }}>{details}</Text>
          </Column>
        )}

        {/* Address and Contact in one line */}
        {contactInfo && (
          <Column sm={4} md={8} lg={16}>
            <Text>{contactInfo}</Text>
          </Column>
        )}
      </Grid>
    </Tile>
  );
};

export default PatientDetails;
