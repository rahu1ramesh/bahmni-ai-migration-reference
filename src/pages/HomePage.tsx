import React from 'react';
import { Grid, Column, Section } from '@carbon/react';
import PatientDetails from '@components/patient/PatientDetails';
import ConditionsTable from '@components/conditions/ConditionsTable';
import AllergiesTable from '@components/allergies/AllergiesTable';
import TreatmentsTable from '@components/treatments/TreatmentsTable';

const HomePage: React.FC = () => {
  return (
    <Section>
      <Grid>
        <Column lg={16} md={8} sm={4}>
          <PatientDetails />
          <AllergiesTable />
          <ConditionsTable />
          <TreatmentsTable />
        </Column>
      </Grid>
    </Section>
  );
};

export default HomePage;
