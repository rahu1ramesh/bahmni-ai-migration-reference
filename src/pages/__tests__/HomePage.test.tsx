import React from 'react';
import { render, screen } from '@testing-library/react';
import HomePage from '../HomePage';
import PatientDetails from '@components/patient/PatientDetails';
import ConditionsTable from '@components/conditions/ConditionsTable';

// Mock the PatientDetails component
jest.mock('@components/patient/PatientDetails', () => {
  return jest.fn(() => (
    <div data-testid="mocked-patient-details">Mocked PatientDetails</div>
  ));
});

// Mock the ConditionsTable component
jest.mock('@components/conditions/ConditionsTable', () => {
  return jest.fn(() => (
    <div data-testid="mocked-conditions-table">Mocked ConditionsTable</div>
  ));
});

// Mock the AllergiesTable component
jest.mock('@components/allergies/AllergiesTable', () => {
  return jest.fn(() => (
    <div data-testid="mocked-allergy-table">Mocked AllergiesTable</div>
  ));
});
describe('HomePage Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render without crashing', () => {
    render(<HomePage />);
    expect(screen.getByTestId('mocked-conditions-table')).toBeInTheDocument();
    expect(screen.getByTestId('mocked-patient-details')).toBeInTheDocument();
  });

  it('should render PatientDetails component', () => {
    render(<HomePage />);
    expect(PatientDetails).toHaveBeenCalled();
    expect(screen.getByTestId('mocked-patient-details')).toBeInTheDocument();
    expect(screen.getByText('Mocked PatientDetails')).toBeInTheDocument();

    expect(ConditionsTable).toHaveBeenCalled();
    expect(screen.getByTestId('mocked-conditions-table')).toBeInTheDocument();
    expect(screen.getByText('Mocked ConditionsTable')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
