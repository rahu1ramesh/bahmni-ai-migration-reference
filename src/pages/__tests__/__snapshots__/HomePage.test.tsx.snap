// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component should match snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class="cds--css-grid"
    >
      <div
        class="cds--sm:col-span-4 cds--md:col-span-8 cds--lg:col-span-16 cds--css-grid-column"
      >
        <div
          data-testid="mocked-patient-details"
        >
          Mocked PatientDetails
        </div>
        <div
          data-testid="mocked-allergy-table"
        >
          Mocked AllergiesTable
        </div>
        <div
          data-testid="mocked-conditions-table"
        >
          Mocked ConditionsTable
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
